drivers/modbus_rtu/modbus_rtu.c

```diff
@@ -0,0 +1,469 @@
+static void rx_cb(void *arg, uint8_t data);
+
+static inline int wait_bytes(modbus_rtu_t *modbus, uint8_t num) {
+  msg_t msg;
+  while (modbus->size_buffer < num) {
+    if (xtimer_msg_receive_timeout(&msg, modbus->rx_timeout) < 0) {
+      return -1;
+    }
+  }
+  return 0;
+}
+// from buffer
+static inline void read_address(modbus_rtu_t *modbus) {
+  modbus->msg->addr = ((modbus->buffer[ADD_HI]) << 8) | modbus->buffer[ADD_LO];
+}
+// to buffer
+static inline void write_address(modbus_rtu_t *modbus) {
+  modbus->buffer[ADD_HI] = highByte(modbus->msg->addr);
+  modbus->buffer[ADD_LO] = lowByte(modbus->msg->addr);
+}
+// from buffer
+static inline void read_count(modbus_rtu_t *modbus) {
+  modbus->msg->count = ((modbus->buffer[NB_HI]) << 8) | modbus->buffer[NB_LO];
+}
+// to buffer
+static inline void write_count(modbus_rtu_t *modbus) {
+  modbus->buffer[NB_HI] = highByte(modbus->msg->count);
+  modbus->buffer[NB_LO] = lowByte(modbus->msg->count);
+}
+
+static void copy_bits(uint8_t *dst, uint16_t start_bit_dst,
+                      const uint8_t *src, uint16_t start_bit_src, uint16_t count) {
+  while (count > 0) {
+    div_t s = div(start_bit_src, 8);
+    div_t d = div(start_bit_dst, 8);
+    // printf("dst byte: %d; bit: %d\n", d.quot, d.rem);
+    uint8_t bit = src[s.quot] & (1 << s.rem);
+    // printf("src byte: %d; bit: %d, val: %d\n", s.quot, s.rem, bit);
+    if (bit) {
+      dst[d.quot] |= 1 << (7 - d.rem);
+    } else {
+      dst[d.quot] &= ~(1 << (7 - d.rem));
+    }
+    start_bit_dst++;
+    start_bit_src++;
+    count--;
+  }
+}
+
+int modbus_rtu_init(modbus_rtu_t *modbus, uint32_t baudrate) {
+  int err = uart_init(modbus->uart, baudrate, rx_cb, modbus);
+  if (err != UART_OK) {
+    return err;
+  }
+  if (modbus->pin_rts != GPIO_UNDEF) {
+    gpio_init(modbus->pin_rts, GPIO_OUT);
+    gpio_write(modbus->pin_rts, !modbus->pin_rts_enable);
+  }
+  mutex_init(&(modbus->mutex_buffer));
+  // usec in sec / byte per second * 1.5
+  modbus->rx_timeout = 1000000 / (baudrate / 10) * 1.5;
+
+  modbus->size_buffer = 0;
+  modbus->msg = NULL;
+
+  puts("modbus_init");
+
+  return 0;
+}
+
+int modbus_rtu_send_request(modbus_rtu_t *modbus, modbus_rtu_message_t *message) {
+  modbus->msg = message;
+  modbus->pid = thread_getpid();
+  mutex_lock(&(modbus->mutex_buffer));
+  int res = prepare_request(modbus);
+  if (res != MB_ER_NONE) {
+    mutex_unlock(&(modbus->mutex_buffer));
+    return res;
+  }
+  send(modbus);
+  modbus->size_buffer = 0;
+  mutex_unlock(&(modbus->mutex_buffer));
+
+  return get_response(modbus);
+}
+
+static inline int prepare_request(modbus_rtu_t *modbus) {
+  modbus->buffer[ID] = modbus->msg->id;
+  modbus->buffer[FUNC] = modbus->msg->func;
+  write_address(modbus);
+  uint8_t size;
+  div_t d;
+  switch (modbus->msg->func) {
+  case MB_FC_READ_COILS:
+  case MB_FC_READ_DISCRETE_INPUT:
+  case MB_FC_READ_REGISTERS:
+  case MB_FC_READ_INPUT_REGISTER:
+    write_count(modbus);
+    // id + func + addr + count
```

Comment:
I guess in the end all the comments will be converted to C style so I will stop nitpicking :)